### PR TITLE
Added echo function to OpenAI API server.

### PR DESCRIPTION
**Pull request by wanmok**
The feature implements the echo function for OpenAI API server. It supersedes previous PR #959 on top of #1328. Also, it supports the issue #201. 

For OpenAI API server, the PR makes following modifications:
- Removed error message for requesting echo.
- Modified create_logprobs function to reflect engine output changes.
- Changed the behavior of streaming as in OpenAI's implementation, the number of streamed payloads basically equals to the number of tokens generated times number of choices; whereas in our previous implementation, it would yield another end of streaming empty token. 
- Edge case: added the support for the case of `echo=True` and `max_tokens=0`, which is a valid case in OpenAI APIs. In this case, the SamplingParams.max_tokens would be set to 1 and enable an echo_self flag. The flag is then used in the following post processing to ensure the additionally generated token being removed from the final output.

### CODE REVIEW

**Comment by wanmok** [comment]
@zhuohan123 request to review.

File: vllm/entrypoints/openai/api_server.py
Comment (zhuohan123) [inline vllm/entrypoints/openai/api_server.py:504]: This is to mimic OpenAI server's behavior. Please add it back. When query the server with:
```python
completion = openai.Completion.create(
    model="text-davinci-003", prompt="Say this is a test", n=1, echo=True, best_of=1,
    stream=True, logprobs=2, max_tokens=5)
for c in completion:
    print(c)
```
The reply is:
```
{
  "choices": [
    {
      "finish_reason": null,
      "index": 0,
      "logprobs": {
        "text_offset": [
          0,
          3,
          8,
          11,
          13,
          18
        ],
        "token_logprobs": [
          null,
          -5.8172827,
          -2.688938,
          -1.3986964,
          -3.46766,
          -0.05558832
        ],
        "tokens": [
          "Say",
          " this",
          " is",
          " a",
          " test",
          "\n"
        ],
        "top_logprobs": [
          null,
          {
            "\n": -2.923694,
            "(\"": -3.1542532
          },
          {
            "\n": -0.79770833,
            " is": -2.688938
          },
          {
            " a": -1.3986964,
            " the": -1.8430934
          },
          {
            " test": -3.46766,
            " very": -3.8154793
          },
          {
            "\n": -0.05558832,
            ".": -4.5110774
          }
        ]
      },
      "text": "Say this is a test\n"
    }
  ],
  "created": 1698764284,
  "id": "cmpl-8FkTsfS1sPFzyue9mEKjl87ZtPmt7",
  "model": "text-davinci-003",
  "object": "text_completion",
  "warning": "This model version is deprecated. Migrate before January 4, 2024 to avoid disruption of service. Learn more https://platform.openai.com/docs/deprecations"
}
{
  "choices": [
    {
      "finish_reason": null,
      "index": 0,
      "logprobs": {
        "text_offset": [
          19
        ],
        "token_logprobs": [
          -0.01684946
        ],
        "tokens": [
          "\n"
        ],
        "top_logprobs": [
          {
            "\n": -0.01684946,
            "Say": -4.6857796
          }
        ]
      },
      "text": "\n"
    }
  ],
  "created": 1698764284,
  "id": "cmpl-8FkTsfS1sPFzyue9mEKjl87ZtPmt7",
  "model": "text-davinci-003",
  "object": "text_completion",
  "warning": "This model version is deprecated. Migrate before January 4, 2024 to avoid disruption of service. Learn more https://platform.openai.com/docs/deprecations"
}
{
  "choices": [
    {
      "finish_reason": null,
      "index": 0,
      "logprobs": {
        "text_offset": [
          20
        ],
        "token_logprobs": [
          -0.54426545
        ],
        "tokens": [
          "This"
        ],
        "top_logprobs": [
          {
            "This": -0.54426545,
            "Yes": -0.95614594
          }
        ]
      },
      "text": "This"
    }
  ],
  "created": 1698764284,
  "id": "cmpl-8FkTsfS1sPFzyue9mEKjl87ZtPmt7",
  "model": "text-davinci-003",
  "object": "text_completion",
  "warning": "This model version is deprecated. Migrate before January 4, 2024 to avoid disruption of service. Learn more https://platform.openai.com/docs/deprecations"
}
{
  "choices": [
    {
      "finish_reason": null,
      "index": 0,
      "logprobs": {
        "text_offset": [
          24
        ],
        "token_logprobs": [
          -0.047382396
        ],
        "tokens": [
          " is"
        ],
        "top_logprobs": [
          {
            " indeed": -3.6144009,
            " is": -0.047382396
          }
        ]
      },
      "text": " is"
    }
  ],
  "created": 1698764284,
  "id": "cmpl-8FkTsfS1sPFzyue9mEKjl87ZtPmt7",
  "model": "text-davinci-003",
  "object": "text_completion",
  "warning": "This model version is deprecated. Migrate before January 4, 2024 to avoid disruption of service. Learn more https://platform.openai.com/docs/deprecations"
}
{
  "choices": [
    {
      "finish_reason": "length",
      "index": 0,
      "logprobs": {
        "text_offset": [
          27
        ],
        "token_logprobs": [
          -0.028564425
        ],
        "tokens": [
          " indeed"
        ],
        "top_logprobs": [
          {
            " a": -3.7224274,
            " indeed": -0.028564425
          }
        ]
      },
      "text": " indeed"
    }
  ],
  "created": 1698764284,
  "id": "cmpl-8FkTsfS1sPFzyue9mEKjl87ZtPmt7",
  "model": "text-davinci-003",
  "object": "text_completion",
  "warning": "This model version is deprecated. Migrate before January 4, 2024 to avoid disruption of service. Learn more https://platform.openai.com/docs/deprecations"
}
{
  "choices": [
    {
      "finish_reason": "length",
      "index": 0,
      "logprobs": {
        "text_offset": [],
        "token_logprobs": [],
        "tokens": [],
        "top_logprobs": []
      },
      "text": ""
    }
  ],
  "created": 1698764284,
  "id": "cmpl-8FkTsfS1sPFzyue9mEKjl87ZtPmt7",
  "model": "text-davinci-003",
  "object": "text_completion",
  "warning": "This model version is deprecated. Migrate before January 4, 2024 to avoid disruption of service. Learn more https://platform.openai.com/docs/deprecations"
}

```
Note in the very last json reply it only includes the finish reason.

**Review by zhuohan123 — CHANGES_REQUESTED** [review]
Thank you for your contribution! Please check the comments.

File: vllm/entrypoints/openai/api_server.py
Comment (wanmok) [reply vllm/entrypoints/openai/api_server.py:504]: Interesting... It looks like I tried other OpenAI models, and they behaved in a different way. Adding them back.

File: vllm/entrypoints/openai/api_server.py
Comment (wanmok) [reply vllm/entrypoints/openai/api_server.py:211]: This is actually different from the implementation. Run:
```
completion = openai.Completion.create(
    model="text-davinci-003", prompt="A robot may not injure a human being", n=1, echo=True, best_of=1,
    logprobs=2, max_tokens=5, temperature=2)
```

Response:
```
/Users/yunmochen/anaconda3/envs/vllm/bin/python /Users/yunmochen/research/vllm/examples/openai_completion_client.py 
Completion results:
{
  "warning": "This model version is deprecated. Migrate before January 4, 2024 to avoid disruption of service. Learn more https://platform.openai.com/docs/deprecations",
  "id": "cmpl-8Ft6vvaOBBcbTtdxF3P4pLX1NrP58",
  "object": "text_completion",
  "created": 1698797457,
  "model": "text-davinci-003",
  "choices": [
    {
      "text": "A robot may not injure a human being or muddy pave collectorCy",
      "index": 0,
      "logprobs": {
        "tokens": [
          "A",
          " robot",
          " may",
          " not",
          " injure",
          " a",
          " human",
          " being",
          " or",
          " muddy",
          " pave",
          " collector",
          "Cy"
        ],
        "token_logprobs": [
          null,
          -10.330451,
          -0.605383,
          -3.694869,
          -0.12196065,
          -0.007913817,
          -0.0015808053,
          -0.0075414344,
          -0.049919397,
          -19.736057,
          -11.607917,
          -18.587305,
          -14.40027
        ],
        "top_logprobs": [
          null,
          {
            "pi": -3.0112953,
            ".": -3.0572982
          },
          {
            " may": -0.605383,
            " must": -2.4434366
          },
          {
            " be": -0.22049057,
            " perform": -3.1058216
          },
          {
            " injure": -0.12196065,
            " harm": -2.4350166
          },
          {
            " a": -0.007913817,
            " or": -6.1911945
          },
          {
            " human": -0.0015808053,
            "\n": -7.691319
          },
          {
            " being": -0.0075414344,
            " or": -6.095392
          },
          {
            " or": -0.049919397,
            ",": -3.4204276
          },
          {
            ",": -0.013517476,
            " through": -4.427107
          },
          {
            " through": -1.5732663,
            " a": -1.6822541
          },
          {
            "ments": -0.031058038,
            " the": -5.4342465
          },
          {
            ",": -0.95207053,
            " through": -1.8842442
          }
        ],
        "text_offset": [
          0,
          1,
          7,
          11,
          15,
          22,
          24,
          30,
          36,
          39,
          45,
          50,
          60
        ]
      },
      "finish_reason": "length"
    }
  ],
  "usage": {
    "prompt_tokens": 8,
    "completion_tokens": 5,
    "total_tokens": 13
  }
}
```

The API guarantees the log probabilities of chosen tokens present in `token_logprobs` but not `top_logprobs`.

**Comment by wanmok** [comment]
@zhuohan123 Except for the unresolved conversation, I've addressed comments.

File: vllm/entrypoints/openai/api_server.py
Comment (tmostak) [reply vllm/entrypoints/openai/api_server.py:211]: Just as another datapoint, the `llama-cpp-python` OpenAI-spec server also includes the probabilities of chosen tokens in `top_logprobs`.

Comment (wanmok) [reply vllm/entrypoints/openai/api_server.py:211]: It's just about the decision whether we would like to follow the service behavior implemented by OpenAI or the documented specifications.

**Comment by wanmok** [comment]
@zhuohan123 besides the design decision for the top logprob behavior, is there anything you would like to address?

**Comment by lifengjin** [comment]
any updates?

**Comment by wanmok** [comment]
> any updates?

I suppose that we’re still waiting for the team to get another round of review done. @zhuohan123 do we have any concern with merging the PR?

**Comment by tmostak** [comment]
@zhuohan123 would it be possible to get this PR re-reviewed/merged... we're heavily using this branch but would be much better if the code was in main. 

Haven't reviewed the code but all seems to work well (aside from the open question on whether the logprobs of the chosen tokens should be returned in `top_logprobs`).

File: vllm/entrypoints/openai/api_server.py
Comment (zhuohan123) [reply vllm/entrypoints/openai/api_server.py:217]: Let's follow the OpenAI's API description and revert the changes here?

File: vllm/entrypoints/openai/api_server.py
Comment (zhuohan123) [inline vllm/entrypoints/openai/api_server.py:448]: BTW are you using a different formatter from the `format.sh` in the project root directory? This PR includes many unnecessary style changes like this one. Before we merge, can you minimize these unnecessary style change? Thanks.

File: vllm/entrypoints/openai/api_server.py
Comment (zhuohan123) [inline vllm/entrypoints/openai/api_server.py:183]: Using `PromptLogporbs` here is confusing since this function is not only for prompt log probs. Let's just use `List[Optional[Dict[int, float]]]` here for clarity?

File: vllm/entrypoints/openai/protocol.py
Comment (zhuohan123) [inline vllm/entrypoints/openai/protocol.py:110]: Same, just use `List[Optional[Dict[int, float]]]` here.

File: vllm/entrypoints/openai/api_server.py
Comment (zhuohan123) [inline vllm/entrypoints/openai/api_server.py:40]: Again, this is an unnecessary style change.

**Review by zhuohan123 — CHANGES_REQUESTED** [review]
Hi thanks for your contributions again! I believe after these changes the PR should be able to be merged. Could you remove all the other unnecessary style changes in this PR? Thanks

File: vllm/entrypoints/openai/api_server.py
Comment (wanmok) [reply vllm/entrypoints/openai/api_server.py:446]: I used the `format.sh`, which then introduces multiple changes... I suspect that it might result from some formatter version mismatch.

**Comment by wanmok** [comment]
@zhuohan123 I've addressed the comments. Please let me know if you found anything else to address.

**Comment by zhuohan123** [comment]
Hi thanks for fixing the issues! I tried running the following script to test:
```python
import openai

# Modify OpenAI's API key and API base to use vLLM's API server.
openai.api_key = "EMPTY"
openai.api_base = "http://localhost:8000/v1"

# List models API
models = openai.Model.list()
print("Models:", models)

model = models["data"][0]["id"]

# Completion API
stream = False
completion = openai.Completion.create(
    model=model,
    prompt="A robot may not injure a human being",
    echo=True,
    n=2,
    stream=stream,
    logprobs=3)

print("Completion results:")
if stream:
    for c in completion:
        print(c)
else:
    print(completion)
```

And got the following error on the server side:
```
Traceback (most recent call last):
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 426, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/fastapi/applications.py", line 289, in __call__
    await super().__call__(scope, receive, send)
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/middleware/cors.py", line 83, in __call__
    await self.app(scope, receive, send)
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
    raise exc
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 20, in __call__
    raise e
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 17, in __call__
    await self.app(scope, receive, send)
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/routing.py", line 66, in app
    response = await func(request)
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/fastapi/routing.py", line 273, in app
    raw_response = await run_endpoint_function(
  File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/fastapi/routing.py", line 190, in run_endpoint_function
    return await dependant.call(**values)
  File "/home/zhuohan/vllm/vllm/vllm/entrypoints/openai/api_server.py", line 594, in create_completion
    logprobs = create_logprobs(
  File "/home/zhuohan/vllm/vllm/vllm/entrypoints/openai/api_server.py", line 193, in create_logprobs
    for i, p in step_top_logprobs.items()
AttributeError: 'NoneType' object has no attribute 'items'
```

**Comment by wanmok** [comment]
> Hi thanks for fixing the issues! I tried running the following script to test:
> 
> ```python
> import openai
> 
> # Modify OpenAI's API key and API base to use vLLM's API server.
> openai.api_key = "EMPTY"
> openai.api_base = "http://localhost:8000/v1"
> 
> # List models API
> models = openai.Model.list()
> print("Models:", models)
> 
> model = models["data"][0]["id"]
> 
> # Completion API
> stream = False
> completion = openai.Completion.create(
>     model=model,
>     prompt="A robot may not injure a human being",
>     echo=True,
>     n=2,
>     stream=stream,
>     logprobs=3)
> 
> print("Completion results:")
> if stream:
>     for c in completion:
>         print(c)
> else:
>     print(completion)
> ```
> 
> And got the following error on the server side:
> 
> ```
> Traceback (most recent call last):
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 426, in run_asgi
>     result = await app(  # type: ignore[func-returns-value]
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
>     return await self.app(scope, receive, send)
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/fastapi/applications.py", line 289, in __call__
>     await super().__call__(scope, receive, send)
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/applications.py", line 122, in __call__
>     await self.middleware_stack(scope, receive, send)
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
>     raise exc
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
>     await self.app(scope, receive, _send)
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/middleware/cors.py", line 83, in __call__
>     await self.app(scope, receive, send)
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
>     raise exc
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
>     await self.app(scope, receive, sender)
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 20, in __call__
>     raise e
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 17, in __call__
>     await self.app(scope, receive, send)
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/routing.py", line 718, in __call__
>     await route.handle(scope, receive, send)
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/routing.py", line 276, in handle
>     await self.app(scope, receive, send)
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/starlette/routing.py", line 66, in app
>     response = await func(request)
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/fastapi/routing.py", line 273, in app
>     raw_response = await run_endpoint_function(
>   File "/home/zhuohan/anaconda3/envs/vllm/lib/python3.10/site-packages/fastapi/routing.py", line 190, in run_endpoint_function
>     return await dependant.call(**values)
>   File "/home/zhuohan/vllm/vllm/vllm/entrypoints/openai/api_server.py", line 594, in create_completion
>     logprobs = create_logprobs(
>   File "/home/zhuohan/vllm/vllm/vllm/entrypoints/openai/api_server.py", line 193, in create_logprobs
>     for i, p in step_top_logprobs.items()
> AttributeError: 'NoneType' object has no attribute 'items'
> ```
> 
> @wanmok Can you double check on your side? I believe the bug is caused by that the first token of the prompt does not have log probability, and thus its `step_top_logprobs` will be None.

I have it fixed and tested. Please let me know if you found any other issues.

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thank you for your contribution!

**Comment by zhuohan123** [comment]
> does this work? I'm getting "echo is not currently supported" @zhuohan123 @wanmok

Are you using the latest main branch? This PR hasn't been included in the latest release yet.

**Comment by lifengjin** [comment]
When will this be released? I work in a cuda 11.8 environment, and I haven't figured out a way to compile the repo.

**Comment by robertsw22** [comment]
I'm having the same problem as @lifengjin. How do you get a pull request to compile? I'm running into an abi compiler incompatibility error @zhuohan123 .

**Pull request merged**
